celery/security/serialization.py

```diff
@@ -55,37 +55,20 @@ def deserialize(self, data):
     def _pack(self, body, content_type, content_encoding, signer, signature,
               sep=str_to_bytes('\x00\x01')):
         fields = sep.join(
-            ensure_bytes(s) for s in [signer, signature, content_type,
-                                      content_encoding, body]
+            ensure_bytes(s) for s in [b64encode(signer), b64encode(signature),
+                                      content_type, content_encoding, body]
         )
         return b64encode(fields)
 
     def _unpack(self, payload, sep=str_to_bytes('\x00\x01')):
         raw_payload = b64decode(ensure_bytes(payload))
-        first_sep = raw_payload.find(sep)
```

Comment:
I would like to know bit more detail about the changes proposed, care to provide bit more?

Reply:
Sure @auvipy . This is the logic that we have today for unpacking the serialized data:
1. extract signer and `signature_start_index` using first appearance of `sep`
2. get signature bytes according to `signature_start_index` and signature length. Note, that we assume. here that the signature is decoded as-is in the raw_payload, so we can use the signature length to calculate `signature_end_index`
3. extract all other fields from remaining bytes using split by `sep`

The proposed solution simplifies this process, by assuming that the `sep` bytes can be present in the payload only as separators (and not randomly be as part of signature bytes, for example). Since we guarantee that when we encode the binary data using `base64` (which will not contain `\x00\x01`).
With that assumption, we can just split all fields according to `sep` and get all fields at once.
The only exception here is the `body` itself, which may contain the `sep` bytes inside. Since the body is encoded to the end of the payload, we can limit the number of splits according to `sep`, to prevent more splits in that case.
(We can also encode the body as base64 and then we can assure `sep` will not be in `body` as well, but I like this better since it is more performant).